### PR TITLE
Make type of type hole available in quoted patterns

### PR DESCRIPTION
**Pull request by nicolasstucki**
### Current state
In general, instances of `scala.quoted.Type` are and should not be used explicitly in quotes.

```diff
+ def f[T](x: T)(using Type[T])(using QuoteContext) = 
+   '{ val y: T = $x; ... } // good code

- def f[T](x: T)(using t: Type[T])(using QuoteContext) = 
-   '{ val y: t.T = $x; ... } // bad code

- def f[T](x: T)(using t: Type[T])(using QuoteContext) = 
-   '{ val y: $t = $x; ... } // bad code (syntax will be removed)
```

The only exception is on types extracted using quoted patterns as these hide tier type inside an instance of `Type[T]` which must be accessed using it's path.
```diff
??? match 
  case '{ $x: $t } =>
-   '{ val y: t.T = $x; ... } // bad code
-   '{ val y: $t = $x; ... } // bad code (syntax will be removed)
```

Here `t` is provided as a `given` as if we refer to its type we must be able to summon it.

### Changes
The change is to provide the name of the type that is extracted rather than the `given Type[T]`. 
`Type[T]` can be summoned were needed as in all other uses of abstract types in quotes.

```diff
??? match 
  case '{ $x: $T } =>
+   '{ val y: T = $x; ... } // good code
-   val t: Type[T] = summon[Type[T]]
-   '{ val y: t.T = $x; ... } // bad code
-   '{ val y: $t = $x; ... } // bad code (syntax will be removed)
```

The pattern will make `T` and a `given Type[T]` available on the right-hand side of the pattern.
The `T` can be used the same way it could be used in the implementation of a method with signature:
```scala
def f[T](x: T)(using Type[T])(using QuoteContext) = 
  '{ val y: T = $x; ... }
```

```scala
case '{ $x: $T } =>
  '{ val y: T = $x; ... }
```

It is also possible to use lower cases for the type splices `case '{ $x: $t } => '{ val y: t = $x; ... }`.
This is similar to the use of type variables in normal patterns `case Some[t](x) =>`.
It is preferable to use upper cases as it makes it clearer that this is a type and renders better on the uses in the right-hand side of the pattern.

### CODE REVIEW

**Comment by nicolasstucki** [comment]
Note that this was the original design but we went the other way due to technical challenges while typing the quoted patterns. Now I see that it was actually not that complicated to implement this, just had to use semantic names correctly.

**Comment by smarter** [comment]
```scala
 case '{ $x: $T } =>
   '{ val y: T = $x; ... } // good code
```
Isn't it weird that in the pattern we need a dollar before both the term and type identifier, but in the expression it's only for the term? If we're removing the dollar from the expression it sounds like it should also be removed from the pattern:
```scala
 case '{ $x: t } =>
   '{ val y: t = $x; ... } // good code
```

**Comment by nicolasstucki** [comment]
The following have different sematics (which did not change in this PR)
```scala
 case '{ $x: T } =>
 case '{ $x: $T } =>
```
the first states that we are using a type from outside the pattern while the second states the type will be extracted from the pattern and made available. There are use cases for both operations.

**Comment by smarter** [comment]
Right, to avoid the `$` you'd have to use a lowercase `t` for the type to make it clear that it's a pattern, just like in regular type patterns.

**Comment by nicolasstucki** [comment]
How would we lower case-wildcard?

**Comment by nicolasstucki** [comment]
The syntax discussion is out of the topic of this PR. We should have it on a proper issue. I will create one later.

**Comment by nicolasstucki** [comment]
I created #10050 to follow up the syntax discussions

File: compiler/src/dotty/tools/dotc/core/NameOps.scala [hunk 1]
```diff
@@ -80,15 +80,16 @@ object NameOps {
...
+    def isVariableName: Boolean =
+      testSimple { n =>
+        n.length > 0 && {
+          val first = n.head
+          (((first.isLower && first.isLetter) || first == '_')
+            && (n != false_)
+            && (n != true_)
+            && (n != null_))
+        }
+      } || name.is(PatMatQuoteTypeEv)
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/core/NameOps.scala:92]: Why this change is needed? This kind of specialization is not scalable.

File: tests/pos-macros/i6997c.scala [hunk 2]
```diff
@@ -6,9 +6,9 @@ inline def mcr(x: => Any): Any = ${mcrImpl('x)}
 
 def mcrImpl(body: Expr[Any])(using ctx: QuoteContext): Expr[Any] =
   body match
-    case '{$x: $t} =>
+    case '{$x: $T} =>
       '{
-        val tmp: $t = $x
+        val tmp: T = $x
```
Comment (liufengyun) [inline tests/pos-macros/i6997c.scala:11]: This is not as symmetric as before. I think meta-programmers will be at a loss about whether to use `$` or not before a type.

**Review by liufengyun — APPROVED** [review]
Otherwise, LGTM

File: tests/pos-macros/i6997c.scala (hunk 2 above)
Comment (nicolasstucki) [reply tests/pos-macros/i6997c.scala:11]: It is not symmetric because it is not intended to be as quoted types do not follow the same level symmetry that expression do. The syntax `$` for types won't exist outside of patterns and hence they will have a single option and won't be able to get confused about this at the use site. The use of `$` in the pattern is its own issue (see #10050) which we will address separately. Here we focus on the use sites and keep the changes to the patterns to a minimum.

File: compiler/src/dotty/tools/dotc/core/NameOps.scala (hunk 1 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/core/NameOps.scala:92]: For a `$T` we need to create a name while typing for the `given Type[T] ` that will be available if the pattern matched. We need to use a semantic name to make sure that this name does not leak into userspace. This change makes this particular kind of semantic name be a valid pattern var to make typer handle it as such. 

I don't see any other case where we would use something like this and therefore I'm not sure we need it to scale. If we have another use-case in the future we can revise this and create a more general mechanism. We could achieve the same if we use an attachment on the `Ident` to mark such trees but this would come with some extra runtime overhead.

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/core/NameOps.scala:92]: Looking at it, `PatMatQuoteTypeEv` is not needed for anything specific quote wise. I can rename to `VarPattern` or something like that.

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/core/NameOps.scala]: Now it scales to any other synthetic var pattern.

File: compiler/src/dotty/tools/dotc/core/NameOps.scala [hunk 3]
```diff
@@ -79,16 +79,17 @@ object NameOps {
...
+    def isVarPattern: Boolean =
+      testSimple { n =>
+        n.length > 0 && {
+          val first = n.head
+          (((first.isLower && first.isLetter) || first == '_')
+            && (n != false_)
+            && (n != true_)
+            && (n != null_))
+        }
+      } || name.is(PatMatVarName)
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/core/NameOps.scala:92]: The renaming makes everything much more clear 👍

**Review by liufengyun — APPROVED** [review]
LGTM

File: compiler/src/dotty/tools/dotc/core/NameOps.scala (hunk 3 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/core/NameOps.scala:92]: Indeed. The scalability concern made me realize that it was just a misnomer.
